File: C++/322. CoinChange.cpp
"""
Problem Statement:
You are given coins of different denominations and a total amount of money amount. Write a function to compute the fewest number of coins that you need to make up that amount. 
If that amount of money cannot be made up by any combination of the coins, return -1.
You may assume that you have an infinite number of each kind of coin.
Sample Input:
coins = [1,2,5], amount = 11
Output : 3
Explanation : 11 = 5 + 5 + 1
"""
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit=0;
        int n=prices.size();
        if(n==0 || n==1)
            return profit;
        int min_price=prices[0];
        for(int i=0;i<n;i++){
            min_price=min(min_price,prices[i]);
            profit=max(prices[i]-min_price,profit);
        }
        return profit;                    
    }
};

File: Java/322. CoinChange.java
/* You are given coins of different denominations and a total amount of money amount. Write a function to compute the fewest number of coins that you need to make up that amount. If that amount of money cannot be made up by any combination of the coins, return -1.

You may assume that you have an infinite number of each kind of coin.

Example 1:

Input: coins = [1,2,5], amount = 11
Output: 3
Explanation: 11 = 5 + 5 + 1

Example 2:

Input: coins = [2], amount = 3
Output: -1
*/

class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount+1];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] =0;
        
        for(int i=0; i<coins.length; i++){
            for(int j=coins[i]; j<dp.length; j++){
                if(dp[j-coins[i]] != Integer.MAX_VALUE){
                    dp[j] = Math.min(dp[j],dp[j-coins[i]] + 1); 
                }
                
            }
        }
        
        if(dp[amount] != Integer.MAX_VALUE){
            return dp[amount];
        }else{
            return -1;
        }
    }
}
